### PR TITLE
Add a warning when using `encoding` in SSR and `headers` on objects in endpoints

### PR DESCRIPTION
**Pull request by Princesseuh**
## Changes

Title is a mouthful. Basically, `encoding` is only used to write the files when doing static generation, as such in SSR it doesn't do anything, but people expect it to affect the charset, but since https://github.com/withastro/astro/pull/3841 we always default to `charset=utf-8`. In the future, it would be much better if we only returned responses...

The warning for headers is because returning an object with a certain type of body trips up TypeScript into thinking you're returning a Response since they're "compatible", so a small warning will help nudge users in the right direction.

Close https://github.com/withastro/astro/pull/3841

## Testing

N/A. Tested the warning manually

## Docs

N/A

### CODE REVIEW

**Review by ematipico — APPROVED** [review]
Just a nitpick suggestion

**Review by matthewp — APPROVED** [review]

**Review by natemoo-re — CHANGES_REQUESTED** [review]
Should account for `prerender` as well

File: packages/astro/src/core/endpoint/index.ts [hunk 1]
```diff
@@ -111,6 +113,24 @@ export async function call(
 		};
 	}
 
+	if (env.ssr) {
```
[the remark this answers is not in the document]
Comment (natemoo-re) [reply packages/astro/src/core/endpoint/index.ts:116]:
```suggestion
	if (env.ssr && !mod.prerender) {
```

File: packages/astro/src/core/endpoint/index.ts [hunk 2]
```diff
@@ -111,6 +113,24 @@ export async function call(
...
+		if (response.hasOwnProperty('headers')) {
+			warn(
+				logging,
+				'ssr',
+				'Setting headers is not supported when returning an object. Please return an instance of Response. See https://docs.astro.build/en/core-concepts/endpoints/#server-endpoints-api-routes for more information.'
+			);
+		}
+
+		if (response.encoding && !mod.prerender) {
```
Comment (natemoo-re) [inline packages/astro/src/core/endpoint/index.ts:125]:
```suggestion
		if (response.encoding) {
```

**Review by natemoo-re — CHANGES_REQUESTED** [review]
The check should apply to both `headers` and `encoding` I think?

**Review by natemoo-re — APPROVED** [review]
Looks great!

**Pull request merged**